beehive-lab/TornadoVM
# Fix update reference for COPY_IN objects

#### Description

This PR contains the following changes:
- Clean the Local/GlobalObjectState objects. Remove the concept of `owner` device.
- Fix TaskSchedule::updateReference for objects that are copied via a COPY_IN bytecode. `Resize::testUpdateReferenceCopyIn` tests this scenario.
- Expand profiling information for sync operations and differentiate between dispatch kernel and data transfer times. `TestProfiler::testProfilerEnabled` includes the sync profile info case.

#### Problem description

The problems fixed in this PR are the following:
1) When `TaskSchedule::updateReference` is called on a variable that is copied in (via a `COPY_IN` bytecode), the runtime thinks that the data is already on the device and will not perform any update. This is fixed by invalidating the object state and updating the necessary datastructures. (Check `TornadoExecutionContext::replaceObjectState`)

2) Expand the profiling information to include times for `sync` operations; Include call stack writing in the reported COPY_IN times. Split `DISPATCH_TIME` into `TOTAL_DISPATCH_DATA_TRANSFERS_TIME` and `TOTAL_DISPATCH_KERNEL_TIME`.

3) Remove the GridTask from the `TornadoVM` constructor. Instead, update the `GridTask` of the `TornadoVM` object separately, since the same `TornadoVM` object might be used with different `GridTask`s

#### Backend/s tested


#### OS tested


#### Did you check on FPGAs?

If possible, check your changes on FPGAs.


#### How to test the new patch?

Running the unit tests should cover this PR. 
For the update reference fix, check `Resize::testUpdateReferenceCopyIn`
